File: Ruby/implementation_4/lib/board.rb
require './lib/seeds'

def max_x(pairs)
  pairs.max_by {|pair| pair[0]}[0]
end

def min_x(pairs)
  pairs.min_by {|pair| pair[0]}[0]
end

def max_y(pairs)
  pairs.max_by {|pair| pair[1]}[1]
end

def min_y(pairs)
  pairs.min_by {|pair| pair[1]}[1]
end

def offset(pairs, delta)
  dx = delta[0]
  dy = delta[1]

  pairs.collect do |pair|
    [pair[0]+dx, pair[1]+dy]
  end
end


NEIGHBOR_OFFSET = [[-1,1], [0,1], [1,1],
                   [-1,0],        [1,0],
                   [-1,-1],[0,-1],[1,-1]]



class Board
  attr_reader :x_min, :x_max, :y_min, :y_max, :trail

  def initialize(seed, trail=[])
    if seed.nil?
      raise ArgumentError
    end
    @trail = trail.concat(seed).uniq
    @x_min = min_x(@trail)
    @x_max = max_x(@trail)
    @y_min = min_y(@trail)
    @y_max = max_y(@trail)

    @board = seed.uniq

  end

  def print_grid
    (@y_max).downto(@y_min) do |y_index|
      (@x_min).upto(@x_max) do |x_index|
        print (@board.include?([x_index, y_index]) ? '#' : '+')
        print ' '
      end
      puts
    end
  end

  def next_generation
    next_board = []
    relevant_cells = []
    count_hash = {}

    @board.each do |pair|
      relevant_cells.concat offset(NEIGHBOR_OFFSET, pair)
    end

    element = relevant_cells.pop
    while !element.nil? do
      number = 1
      number += relevant_cells.count(element)
      relevant_cells.delete(element)
      count_hash[element] = number
      element = relevant_cells.pop
    end

    count_hash.each_pair do |key, value|
      if ((value == 3) || ((value == 2) && @board.include?(key)))
        next_board << key
      end
    end

    Board.new(next_board, @trail)
  end

  def column(number)
    @board.select do |cell|
        cell[0] == number
    end
  end

  def row(number)
    @board.select do |cell|
      cell[1] == number
    end
  end

  def width
    @x_max - @x_min + 1
  end

  def height
    @y_max - @y_min + 1
  end

  def live
    @board.length
  end

  def size
    self.width * self.height
  end

  def include?(pair)
    @board.include? pair
  end

  
end

File: JavaScript/js/board.js
function Board(seed, trail) {
  if (!seed) {
    throw new Error;
  }
  this.board = uniq(seed);
  this.trail = uniq(mergeWithoutFlatten(trail, seed));
  this.xMax = maxX(this.trail);
  this.xMin = minX(this.trail);
  this.yMax = maxY(this.trail);
  this.yMin = minY(this.trail);
}

Board.prototype = {
  board: [ ],
  trail: [ ],

  printGrid: function() {
    for (var j = this.yMax; j >= this.yMin; j--) {
      var row = '';
      for (var i = this.xMin; i <= this.xMax; i++) {
        row += (contains(this.board, [i, j]) ? '# ' : '+ ');
      }
      console.log(row + '\n');
    }
  },

  width: function() {
    return this.xMax - this.xMin + 1;
  },

  height: function() {
    return this.yMax - this.yMin + 1;
  },

  size: function() {
    return this.width() * this.height();
  },

  live: function() {
    return this.board.length;
  },

  column: function(number) {
    return this.board.filter(function(element) {
      return element[0] === number;
    });
  },

  row: function(number) {
    return this.board.filter(function(element) {
      return element[1] === number;
    });
  },

  includes: function(pair) {
    return contains(this.board, pair);
  },

  nextGeneration: function() {
    var nextBoard = [ ];
    var relevantCells = [ ];
    var countHash = [ ];

    for (var i = 0; i < this.board.length; i++) {
      var pair = this.board[i];
      relevantCells =  mergeWithoutFlatten(relevantCells, offset(neighborOffset, pair));
    }

    var element = relevantCells.pop();
    while (element) {
      var number = 1;
      number += count(relevantCells, element);
      relevantCells = deleteIf(relevantCells, element, coordinateEquality);
      countHash.push([element, number]);
      element = relevantCells.pop();
    }

    for (var i = 0; i < countHash.length; i++) {
      var key = countHash[i][0];
      var value = countHash[i][1];
      if (value === 3 || (value === 2 && containsCoordinate(this.board, key)))
        nextBoard.push(key);
    }

    return new Board(nextBoard, this.trail);
  }
};


